content/blog/2022/2022-08-30-handling-timezones-in-spring.md

```diff
@@ -0,0 +1,516 @@
+
+For Europe/Berlin at the current date/time :
+
+{{% image alt="settings" src="images/posts/handling-timezones-in-spring/spring_tz_2.JPG" %}}
+{{% image alt="settings" src="images/posts/handling-timezones-in-spring/db_tz_2.JPG" %}}
+
+The results in this timezone are consistent with the points noted above.
+
+Now, lets see what happens at Europe/Berlin when the DST ends on 30th October 2022. **The custom date considered here is 8th November 2022**.
+
+{{% image alt="settings" src="images/posts/handling-timezones-in-spring/spring_tz_3.JPG" %}}
+{{% image alt="settings" src="images/posts/handling-timezones-in-spring/db_tz_3.JPG" %}}
+
+When DST ends, Europe/Berlin will shift to UTC+1 timezone and this is consistent with the results as seen the output above.
+In all above cases, `OffsetDateTime` and `ZonedDateTime` show the same results. This is because, the **`OffsetDateTime` is derived from `ZoneId`**. All DST rules apply to ZoneId
+and hence `OffsetDateTime` gives the correct representation that includes DST changes. As discussed in the API section that details differences between OffsetDateTime and ZonedDateTime,
+we could use the one that best suits our use-case.
+
+## Testing timezones in a Spring Boot application
+
+When working with timezones and unit testing applications, we might want to control the timezone, dates and make them agnostic of the system timezone.
+The Date/Time API provides [`java.time.Clock`](https://docs.oracle.com/javase/8/docs/api/java/time/Clock.html) that can be used for achieving this.
+According to the official documentation:
+> Use of a Clock is optional. All key date-time classes also have a now() factory method that uses the system clock in the default time zone. The primary purpose of this abstraction is to allow alternate clocks to be plugged in as and when required. 
+> Applications use an object to obtain the current time rather than a static method. This can simplify testing.
+
+With this approach, we could define a Clock object in the desired timezone and pass it to any of the Date/Time API classes 
+to get the corresponding date-time.
+
+````java
+@TestConfiguration
+public class ClockConfiguration {
+
+    @Bean
+    public Clock clock() {
+        return Clock.system(ZoneId.of("Europe/London"));
+    }
+}
+````
+
+Now in order to get timezone specific information, we can use
+````java
+OffsetDateTime current = OffsetDateTime.now(clock);
+````
+
+Further, we could also fix the clock to set it to a particular instant in a timezone
+````java
+Clock.fixed(Instant.parse("2022-11-08T09:10:20.00Z"), ZoneId.of("Europe/Berlin"));
+````
+With this set, `OffsetDateTime.now(clock)` will always return the same time.
+
+To always apply the default system timezone, we could use:
+````java
+Clock.systemDefaultZone();
+````
+
+By setting the clock parameter, testing the same application in different timezones, with or without DST becomes much easier.
+
+## Conclusion
```

Comment:
I'm missing a short section "When and how to store timezones in the Database?" that lists the options of how to store timezone information in the database and the pros and cons of each and in which use cases they are best. I.e.:

- store UTC in the database and convert to the user's timezone when reading data out of the DB
- store user timezone in the database (may be required for objects like travel itineraries or manifests or similar things)